src/iface/interface.rs

```diff
@@ -1886,21 +1928,78 @@ impl<'a> InterfaceInner<'a> {
     }
 
     #[cfg(feature = "proto-ipv4")]
-    fn process_ipv4<'frame, T: AsRef<[u8]> + ?Sized>(
+    fn process_ipv4<'output, 'payload: 'output, T: AsRef<[u8]> + ?Sized>(
         &mut self,
         sockets: &mut SocketSet,
-        ipv4_packet: &Ipv4Packet<&'frame T>,
-    ) -> Option<IpPacket<'frame>> {
+        ipv4_packet: &Ipv4Packet<&'payload T>,
+        fragments: &'output mut PacketAssemblerSet<'a, Ipv4FragKey>,
+    ) -> Option<IpPacket<'output>> {
         let ipv4_repr = check!(Ipv4Repr::parse(ipv4_packet, &self.caps.checksum));
-
         if !self.is_unicast_v4(ipv4_repr.src_addr) {
             // Discard packets with non-unicast source addresses.
             net_debug!("non-unicast source address");
             return None;
         }
 
+        check!(fragments.remove_when(|frag| Ok(self.now - frag.start_time()? > frag.ttl()?)));
+        fragments.remove_discarded();
+
+        let ip_payload = {
+            if ipv4_packet.more_frags() || ipv4_packet.frag_offset() != 0 {
+                let key = ipv4_packet.get_key();
+
+                let f = match fragments.get_packet_assembler_mut(&key) {
+                    Ok(f) => f,
+                    Err(_) => {
+                        check!(check!(fragments.reserve_with_key(&key)).start(
+                            ipv4_packet.total_len() as usize - ipv4_packet.header_len() as usize,
+                            self.now,
+                            Duration::from_secs(15),
+                            0,
+                        ));
+                        check!(fragments.get_packet_assembler_mut(&key))
+                    }
+                };
+
+                if ipv4_packet.frag_offset() != 0 {
+                    check!(f.extend_at(
+                        ipv4_packet.frag_offset() as usize,
+                        ipv4_packet.total_len() as usize - ipv4_packet.header_len() as usize
+                    ));
+                    check!(f.update_ttl(
+                        check!(f.ttl()).max(Duration::from_secs(ipv4_packet.hop_limit() as u64)),
```

Comment:
I believe using the hop limit (TTL) as a timeout in seconds is deprecated. See RFC1122 section 3.2.2:

```
There MUST be a reassembly timeout.  The reassembly timeout
value SHOULD be a fixed value, not set from the remaining TTL.
It is recommended that the value lie between 60 seconds and 120
seconds.  If this timeout expires, the partially-reassembled
datagram MUST be discarded and an ICMP Time Exceeded message
sent to the source host (if fragment zero has been received).

DISCUSSION:
The IP specification says that the reassembly timeout
should be the remaining TTL from the IP header, but this
does not work well because gateways generally treat TTL as
a simple hop count rather than an elapsed time.  If the
reassembly timeout is too small, datagrams will be
discarded unnecessarily, and communication may fail.  The
timeout needs to be at least as large as the typical
maximum delay across the Internet.  A realistic minimum
reassembly timeout would be 60 seconds.
```

Reply:
I think not sending a `TimeExceeded` is totally not a problem. TCP would have noticed it already I think, and when you use UDP, then you know it cannot always arrive at the destination. And I don't think other protocols have big payloads.

I can indeed move the check to `poll()`, but this will have a performance hit when we have a lot of fragments not belonging to the same packet (though it is very unlikely this will happen).